### DISCUSSION TITLE
Why do some things need to run at different rates? (Read sensors, Update attitude Control attitude, Motor power distribution.)

### DISCUSSION THREAD

**Discussion by HelloIDTC** [bitcraze/discussions#1335] [Q&A] [answered]
Read sensors at full rate (1000Hz);
Update attitude at 250Hz;
Control attitude(attitude controller) at 500Hz;
Motor power distribution at 1000Hz.

Whether it can run at the same rate?

**Comment by tobbeanton** [comment]
Can you perhaps share a bit more why you would like to run them at the same rate? Some documentation about the controllers you can find [here](https://www.bitcraze.io/documentation/repository/crazyflie-firmware/master/functional-areas/sensor-to-control/controllers/).

One reason to run them at different rates is to save computational power.

**Reply by HelloIDTC** [reply]
Running at the same rate this seems reasonable and intuitive. 
More importantly, if everything runs on the same rate(500Hz or 1000Hz), does that make the control better and more stable? 
I think if the read sensor runs at 1000Hz and the update attitude runs at 250Hz,  is there 3 cycles of sensor data discarded? (Run an attitude update after reading the sensor 4 times)  Similarly, the running rate of the controller is 500Hz, so is there 1 cycle of attitude data discarded?

**Reply by knmcguire** [reply, marked as answer]
Yes that is the case, but you shouldn't forget the purpose of the extended kalman filter either, as that does input the sensor data at a higher rate but filters them out. So the sensor data is not thrown away, but used to create a better state estimate: https://www.bitcraze.io/documentation/repository/crazyflie-firmware/master/functional-areas/sensor-to-control/state_estimators/

Like tobias said, we can't run everything at a 1000 hertz and for position control that will actually not improve the performance. Also, the motors on the crazyflie are limited so there is also a hardware limit there. That said, the firmware is open so if some things doesn't seem right, you can try it out to run some  controllers at a higher rate. There should be defines for those rates so you can adjust them accordingly (and fly safely).

**Reply by HelloIDTC** [reply]
OK. In addition, I would like to know how these values of running rate are determined? Is it determined by considering factors such as computing resources and hardware limitations, and after multiple flights? Perhaps these values are not the best, and I can try to modify them to achieve the desired results.
But the attitude controller runs at 500Hz and the update attitude runs at 250Hz, so the input to the attitude controller is the same attitude data in every 2 cycles?

**Reply by knmcguire** [reply]
seems so! But also mind that the rate attitude controller uses direct gyroscope input so that also goes into play.

So as you noticed, we don't experiment with the controller that much since it just works, and a lot of the design process back in the day was a combo of trial/error and some thought into the process. So there is not a lot of precise answers that we can give on your questions because that. But you are more than welcome to try things out :) just let us know on the results as that will be interesting to us as well
